configs/version_constraint.go

```diff
@@ -24,6 +24,15 @@ func decodeVersionConstraint(attr *hcl.Attribute) (VersionConstraint, hcl.Diagno
 	}
 
 	val, diags := attr.Expr.Value(nil)
+	if !val.IsWhollyKnown() {
```

Comment:
I think the problem here is that we weren't checking whether this `Value` call succeeded, only that the `convert.Convert` call below succeeded. When `Value` totally fails it returns `cty.DynamicVal` as a placeholder value along with the error diagnostics. `convert.Convert` allows converting `cty.DynamicVal` to an unknown value of any type (because it represents an unknown value of an _unknown_ type), so the version call would succeed and produce `cty.UnknownVal(cty.String)`, and thus hit this panic.

In this case though, I think it's fine to just stop here if the evaluation failed, so an explicit check for errors should avoid the panic without the extra error message:

```go
    if diags.HasErrors() {
        return ret, diags
    }
```

HCL has a design constraint that if you don't put any unknown values _in_ then it won't give you any out as long as no errors are encountered. Because we are passing `nil` for the `EvalContext` here we are effectively passing no unknown values in (variables and functions are the only way they get introduced), and so if the result has no errors then it is guaranteed to be known and so we shouldn't need any further checks here for known-ness.

Reply:
Ah lovely, that's much more elegant. I did not realize that returned errors.